### ISSUE TITLE
Handling agencies that don't have stop codes

### ISSUE THREAD

**Issue by elliottwilliams** [propershark/timetable_cpp#21]
I'm trying to get Timetable working with BART, and I've run into a conceptual issue I'd like your feedback on: How do we identify stations that don't have a stop code?

BART's `stops.txt` looks like this:

```
stop_id,stop_name,stop_desc,stop_lat,stop_lon,zone_id,stop_url,location_type,parent_station,stop_timezone,wheelchair_boarding
12TH,12th St. Oakland City Center,,37.803768,-122.271450,12TH,http://www.bart.gov/stations/12TH/,0,,,1
16TH,16th St. Mission,,37.765062,-122.419694,16TH,http://www.bart.gov/stations/16TH/,0,,,1
19TH,19th St. Oakland,,37.808350,-122.268602,19TH,http://www.bart.gov/stations/19TH/,0,,,1
...
```

There's no stop code, only a `stop_id`. Looking at the spec for stops.txt, I find this:

> stop_code - Optional	
> Contains short text or a number that uniquely identifies the stop for passengers. Stop codes are often used in phone-based transit information systems or printed on stop signage to make it easier for riders to get a stop schedule or real-time arrival information for a particular stop.

> The stop_code field should only be used for stop codes that are displayed to passengers. For internal codes, use stop_id. This field should be left blank for stops without a code.

Makes sense. BART doesn't have this IRL: Stations are always identified by a full name. Since it's a subway rail system, there are relatively few stops, and they're easy to identify by the street(s) they span or their position in a city.

So why do we key our models with `stop_code`? Looking back at Citybus, I see why: 

```
stop_id,stop_code,stop_name,stop_desc,stop_lat,stop_lon,location_type
73c3bfbe-e120-4cde-ae8a-dc1c661bbe75,"BUS215","CityBus Center: BUS215","CityBus Center",40.4206937802252,-86.8948909279639,0
84f8d586-2af9-498b-b0cd-7c594effe124,"BUS287","2nd St & Columbia St (NW Corner): BUS287","2nd St. and Columbia St. (NW Corner)",40.41832136621,-86.8954126176484,0
```

Presumably, that UUID is something internal to the GTFS file, and doesn't correspond to any IDs we get from Citybus's realtime sources.

What do you think the right approach is here? Should we fall back to `stop_id` when `stop_code` is unavailable, or is there a better approach / refactor you can think of?

**Comment by faultyserver** [comment]
In shark, this is [almost](https://github.com/propershark/shark/blob/v1.0.rc/objects/station.rb#L22) kind of [sorta a configuration](https://github.com/propershark/shark/blob/v1.0.rc/config/doublemap_source.rb#L14). In fact, DoubleMap doesn't even provide `stop_code`s for CityBus, but I was able to [patch it in](https://github.com/propershark/shark/blob/v1.0.rc/config/doublemap_source.rb#L7) by looking at the structure of the data coming from from DoubleMap.

This definitely needs to be made configurable in timetable as well to support this use case. As for why `stop_code` was chosen initially, it was just to make the channel topic more human-readable. UUIDs aren't very human-friendly, making it harder to debug things along the way. At this point, though, I'd be okay with switching over to UUIDs for compatibility.

I think the real question is what makes a good, universal identifier? This is probably different between most agencies out there, so it should be configurable, as mentioned above. But, it also needs to be consistent across the different sources of information for a given agency (GTFS, anything realtime, etc.) so that they can work together and use the same identifiers/channel topics.

In short, the right thing to do is let each system configure it's "primary attribute" for creating an identifier, so that any system with any set of sources can agree on an identification schema.

I don't have an answer for this yet.

**Comment by elliottwilliams** [comment]
As of #22, we now fall back to using the stop `id` if a `code` is unavailable.
